### ISSUE TITLE
DOC: sparse docs "matrix" and "array" mixup

### ISSUE THREAD

**Issue by KatMistberg** [scipy/scipy#18669]
### Issue with current documentation:

- The `sparse` module recommends to use the `array` interfaces over `matrix`, but [under "Usage information" of the `sparse` main page docs](https://scipy.github.io/devdocs/reference/sparse.html#usage-information) all the docs and examples still use the `matrix` versions of the sparse matrices/arrays. I would have expected them to use "array" since that is what is recommended.
- On the other hand, the docs for the individual `matrix` types (e.g. [`bsr_matrix`](https://scipy.github.io/devdocs/reference/generated/scipy.sparse.bsr_matrix.html)) use "array" (e.g. "bsr_array") instead of "matrix". I would have expected `matrix` classes to have "matrix" in the docs, as `array` classes have "array" in the docs. This seems to me to be unintended, since in the code the function `_array_doc_to_matrix` is applied to the array docs to try to swap all instances of "array" for "matrix", but looking at the [definition of the function](https://github.com/scipy/scipy/blob/main/scipy/sparse/_matrix.py), it only replaces "array(s)" with "matrix(ces)" if there is a "sparse" before it.

Apologies if how the current doc is is intended or if these two points should have been made into separate issues.

### Idea or request for content:

- Replace all instances of "matrix" with "array" under "Usage information" ([relevant file](https://github.com/scipy/scipy/blob/main/scipy/sparse/__init__.py#L152))
- Remove "sparse " in the strings in `docstr.replace()` (e.g. "sparse array" -> "array") in the definition of the `_array_doc_to_matrix` function ([relevant file](https://github.com/scipy/scipy/blob/main/scipy/sparse/_matrix.py#L142))

**Comment by dschult** [comment]
I agree that there are some improvements that could be made to the doc_string conversion. But it is not as simple as just changing "array" to "matrix". The doc_strings that refer to `ndarray` or `sparray` or `.toarray` pretty much need to be made into special cases. I'm not sure if there are other changes to avoid. 

Maybe something like:  
```python
import re
re.sub(r"([a-z][a-z][a-z])_array", r"\1_matrix", docs)
```
would help convert the `bsr_array` to `bsr_matrix`...

**Comment by KatMistberg** [comment]
I gave a shot at fixing the 2nd point of the issue (see commit above). I ended up hand-replacing the relevant "array" or "matrix" strings in the docs of the base classes with "{array|matrix}" (and similarly for the plural forms and a/an), which are replaced with "array" or "matrix" in the corresponding array/matrix classes. This method should be more flexible (the instances of "array"/"matrix" can be judged by a case-by-case basis) and easier to maintain (when the docs are edited, this will still work) than the regex solution proposed above.

One thing I noticed though is that this doesn't fix the issue for the docstrings of the _class methods_ (e.g. the docstring of `coo_array.tobsr()` uses "array" while `coo_array.tocsc()` uses "matrix"). To fix this you might have to iterate through all methods and do the replacement. I'll try to tackle this and the 1st point of the issue.

**Comment by dschult** [comment]
These commits look very promising to me. Getting the docs done well is going to be helpful for this transition.
